File: scripts/kitchen-listener.mjs
#!/usr/bin/env node
/**
 * Kitchen Listener - Node.js version (khong can Python)
 * Lang nghe Firestore, tu dong in phieu bep khi co don moi.
 *
 * Can cai 1 lan: npm install firebase-admin
 * Can file:      scripts/serviceAccountKey.json
 * Chay:          node scripts/kitchen-listener.mjs
 */

import { createConnection } from 'net';
import { readFileSync, existsSync } from 'fs';
import { fileURLToPath } from 'url';
import { dirname, join } from 'path';

const __dirname = dirname(fileURLToPath(import.meta.url));

// ── Cau hinh ─────────────────────────────────────────────────────────────────
const PRINTER_IP   = '192.168.1.234';
const PRINTER_PORT = 9100;
const KEY_FILE     = join(__dirname, 'serviceAccountKey.json');
const USE_ASCII    = true;    // ASCII khong dau - tranh ky tu loi tren may in nhiet

// ── ESC/POS constants ─────────────────────────────────────────────────────────
const ESC = 0x1B;
const GS  = 0x1D;
const LF  = 0x0A;
const LINE_WIDTH = 48;

// ── Helper ────────────────────────────────────────────────────────────────────
function log(msg) {
  const t = new Date().toLocaleTimeString('vi-VN');
  console.log(`[${t}] ${msg}`);
}

const VI_MAP = {
  à:'a',á:'a',ả:'a',ã:'a',ạ:'a',
  ă:'a',ằ:'a',ắ:'a',ẳ:'a',ẵ:'a',ặ:'a',
  â:'a',ầ:'a',ấ:'a',ẩ:'a',ẫ:'a',ậ:'a',
  è:'e',é:'e',ẻ:'e',ẽ:'e',ẹ:'e',
  ê:'e',ề:'e',ế:'e',ể:'e',ễ:'e',ệ:'e',
  ì:'i',í:'i',ỉ:'i',ĩ:'i',ị:'i',
  ò:'o',ó:'o',ỏ:'o',õ:'o',ọ:'o',
  ô:'o',ồ:'o',ố:'o',ổ:'o',ỗ:'o',ộ:'o',
  ơ:'o',ờ:'o',ớ:'o',ở:'o',ỡ:'o',ợ:'o',
  ù:'u',ú:'u',ủ:'u',ũ:'u',ụ:'u',
  ư:'u',ừ:'u',ứ:'u',ử:'u',ữ:'u',ự:'u',
  ỳ:'y',ý:'y',ỷ:'y',ỹ:'y',ỵ:'y',đ:'d',
  À:'A',Á:'A',Ả:'A',Ã:'A',Ạ:'A',
  Ă:'A',Ằ:'A',Ắ:'A',Ẳ:'A',Ẵ:'A',Ặ:'A',
  Â:'A',Ầ:'A',Ấ:'A',Ẩ:'A',Ẫ:'A',Ậ:'A',
  È:'E',É:'E',Ẻ:'E',Ẽ:'E',Ẹ:'E',
  Ê:'E',Ề:'E',Ế:'E',Ể:'E',Ễ:'E',Ệ:'E',
  Ì:'I',Í:'I',Ỉ:'I',Ĩ:'I',Ị:'I',
  Ò:'O',Ó:'O',Ỏ:'O',Õ:'O',Ọ:'O',
  Ô:'O',Ồ:'O',Ố:'O',Ổ:'O',Ỗ:'O',Ộ:'O',
  Ơ:'O',Ờ:'O',Ớ:'O',Ở:'O',Ỡ:'O',Ợ:'O',
  Ù:'U',Ú:'U',Ủ:'U',Ũ:'U',Ụ:'U',
  Ư:'U',Ừ:'U',Ứ:'U',Ử:'U',Ữ:'U',Ự:'U',
  Ỳ:'Y',Ý:'Y',Ỷ:'Y',Ỹ:'Y',Ỵ:'Y',Đ:'D',
};
function toAscii(str) {
  return String(str).split('').map(c => VI_MAP[c] ?? c).join('');
}

function enc(str) {
  if (USE_ASCII) str = toAscii(str);
  return [...Buffer.from(String(str), 'utf8')];
}

function fmtTime(ts) {
  if (!ts) return new Date().toLocaleString('vi-VN', { hour:'2-digit', minute:'2-digit', day:'2-digit', month:'2-digit' });
  try {
    let d;
    if (ts.toDate) d = ts.toDate();                        // Firestore Timestamp
    else if (ts._seconds) d = new Date(ts._seconds * 1000); // Firestore Timestamp plain
    else d = new Date(ts);                                   // ISO string hoac so ms
    return d.toLocaleString('vi-VN', { hour:'2-digit', minute:'2-digit', day:'2-digit', month:'2-digit' });
  } catch { return ''; }
}

// ── ESC/POS: tao buffer phieu bep 1 mon ──────────────────────────────────────
function buildKitchenBuffer(tableLabel, time, itemName, qty) {
  const b = [];
  const add  = (...bytes) => b.push(...bytes);
  const str  = s => b.push(...enc(s));
  const line = (s = '') => { str(s); add(LF); };
  const div  = (c = '=') => line(c.repeat(LINE_WIDTH));

  add(ESC, 0x40);                // reset
  add(ESC, 0x61, 0x00);          // can trai

  div();

  // Thoi gian - can giua
  add(ESC, 0x61, 0x01);
  line(time);

  // Ten ban - chu to 2x
  add(GS,  0x21, 0x11);
  add(ESC, 0x45, 0x01);
  line(tableLabel);
  add(GS,  0x21, 0x00);
  add(ESC, 0x45, 0x00);

  add(ESC, 0x61, 0x00);
  line('-'.repeat(LINE_WIDTH));

  // Ten mon - chu cao 2x
  add(GS,  0x21, 0x01);
  add(ESC, 0x45, 0x01);
  line(itemName);
  add(GS,  0x21, 0x00);
  add(ESC, 0x45, 0x00);

  // So luong - can phai, chu to 2x
  add(ESC, 0x61, 0x02);
  add(GS,  0x21, 0x11);
  add(ESC, 0x45, 0x01);
  line(`SL: ${qty}`);
  add(GS,  0x21, 0x00);
  add(ESC, 0x45, 0x00);

  add(ESC, 0x61, 0x00);
  div();

  add(LF, LF);
  add(GS, 0x56, 0x42, 0x00);    // partial cut

  return Buffer.from(b);
}

// ── Gui lenh in qua TCP ───────────────────────────────────────────────────────
function sendToPrinterOnce(buf, timeoutMs = 15000) {
  return new Promise((resolve, reject) => {
    const sock = createConnection({ host: PRINTER_IP, port: PRINTER_PORT }, () => {
      sock.write(buf, err => {
        if (err) { sock.destroy(); reject(err); return; }
        sock.end();
        resolve();
      });
    });
    sock.on('error', reject);
    sock.setTimeout(timeoutMs, () => { sock.destroy(); reject(new Error('timeout')); });
  });
}

async function sendToPrinter(buf) {
  for (let attempt = 1; attempt <= 3; attempt++) {
    try {
      await sendToPrinterOnce(buf);
      return;
    } catch (err) {
      if (attempt === 3) throw err;
      log(`  [RETRY ${attempt}] ${err.message} — thu lai sau 2s...`);
      await new Promise(r => setTimeout(r, 2000));
    }
  }
}

// ── Cache ten mon (ca orderItems lan menuItems) ───────────────────────────────
const nameCache = {};  // id -> ten mon
let db;

async function lookupName(id, collections = ['orderItems', 'menuItems']) {
  if (!id) return null;
  if (nameCache[id]) return nameCache[id];
  for (const col of collections) {
    try {
      const snap = await db.collection(col).doc(id).get();
      if (snap.exists && snap.data().name) {
        nameCache[id] = snap.data().name;
        return nameCache[id];
      }
    } catch {}
  }
  return null;
}

// ── Track items da in ────────────────────────────────────────────────────────
// Key cho mon moi:        "billId:orderItemId"          (in 1 lan)
// Key cho tang so luong:  "billId:orderItemId:delta:addedAt"  (moi lan tang = key moi)
// Key cho custom item moi: "billId:custom:addedAt"      (moi addedAt = key moi)
const printedItems = new Set();

function buildItemKey(billId, item, index) {
  // Custom item moi: dung addedAt lam key de tranh trung voi custom item cu (cung index)
  if (item.customDescription) {
    if (item.addedAt) return `${billId}:custom:${item.addedAt}`;
    return `${billId}:custom:${index}`;
  }
  const id = item.orderItemId || item.menuItemId || index;
  if (item.addedQty > 0 && item.addedAt) {
    return `${billId}:${id}:delta:${item.addedAt}`;
  }
  return `${billId}:${id}`;
}

// ── Xu ly thay doi don hang (ca added lan modified) ──────────────────────────
async function handleBillChange(docSnap, changeType) {
  const billId = docSnap.id;
  const data   = docSnap.data();
  if (!data || data.status !== 'pending') return;

  const isTakeaway = data.isTakeaway || false;
  const tableLabel = isTakeaway
    ? `Mang ve #${data.takeawayNumber}`
    : `Ban ${data.tableNumber}`;
  const items = data.items || [];
  const now   = Date.now();

  let printedCount = 0;
  for (let i = 0; i < items.length; i++) {
    const item    = items[i];

    // Phan biet 3 truong hop:
    //   1. Don moi (added)           : tat ca item, dung createdAt
    //   2. Mon hoan toan moi (modified + addedAt, khong co addedQty)
    //   3. Tang so luong (modified + addedQty + addedAt): key rieng theo addedAt
    const isDelta = changeType !== 'added' && item.addedQty > 0 && item.addedAt;
    const itemKey = buildItemKey(billId, item, i);

    if (printedItems.has(itemKey)) continue;

    // Xac dinh thoi gian de kiem tra "mon cu" (truoc khi listener chay)
    let ts;
    if (changeType === 'added') {
      ts = data.createdAt
        ? (data.createdAt.toDate ? data.createdAt.toDate().getTime() : new Date(data.createdAt).getTime())
        : now;
    } else {
      // ca mon moi lan tang SL deu co addedAt; mon khong doi thi addedAt = undefined -> ts = 0
      ts = item.addedAt ? new Date(item.addedAt).getTime() : 0;
    }

    const ageSec = (now - ts) / 1000;
    if (ageSec > 60) {
      // Mon cu (co truoc khi listener khoi dong) - bo qua, danh dau da xu ly
      printedItems.add(itemKey);
      continue;
    }

    printedItems.add(itemKey);

    // So luong can in:
    // - Tang SL: chi in phan tang them (addedQty), khong in lai ca mon
    // - Mon moi / don moi: in toan bo quantity
    const qty  = isDelta ? item.addedQty : (item.quantity || 1);
    const name = item.name
      || await lookupName(item.orderItemId)
      || await lookupName(item.menuItemId)
      || item.customDescription
      || 'Mon khac';
    const time = changeType === 'added'
      ? fmtTime(data.createdAt)
      : fmtTime(item.addedAt);

    try {
      const buf = buildKitchenBuffer(tableLabel, time, name, qty);
      await sendToPrinter(buf);
      printedCount++;
      log(`  [OK] ${tableLabel} - ${name} x${qty}${isDelta ? ` (them ${qty})` : ''}`);
    } catch (err) {
      log(`  [LOI] ${tableLabel} - ${name} x${qty}: ${err.message}`);
      // Xoa key khoi Set de lan sau thu lai neu dung lai listener
      printedItems.delete(itemKey);
    }
  }

  if (printedCount > 0) {
    log(`Don ${changeType === 'added' ? 'moi' : 'them mon'}: ${tableLabel} - da in ${printedCount} mon`);
  }
}

// ── Main ──────────────────────────────────────────────────────────────────────
async function main() {
  if (!existsSync(KEY_FILE)) {
    console.log(`
[LOI] Khong tim thay: ${KEY_FILE}

Huong dan:
  1. Firebase Console -> Project Settings -> Service accounts
  2. Nhan "Generate new private key" -> tai file JSON
  3. Doi ten thanh serviceAccountKey.json
  4. Dat vao thu muc: ${__dirname}
`);
    process.exit(1);
  }

  // Import firebase-admin (kiem tra da cai chua)
  let admin;
  try {
    admin = (await import('firebase-admin')).default;
  } catch {
    console.log('[LOI] Chua cai firebase-admin. Chay: npm install firebase-admin');
    process.exit(1);
  }

  log('Dang ket noi Firebase...');
  const serviceAccount = JSON.parse(readFileSync(KEY_FILE, 'utf8'));
  admin.initializeApp({ credential: admin.credential.cert(serviceAccount) });
  db = admin.firestore();
  log('Ket noi Firebase thanh cong');

  // Load cache ten mon tu ca orderItems lan menuItems
  const [snapOI, snapMI] = await Promise.all([
    db.collection('orderItems').get(),
    db.collection('menuItems').get(),
  ]);
  snapOI.forEach(d => { if (d.data().name) nameCache[d.id] = d.data().name; });
  snapMI.forEach(d => { if (d.data().name) nameCache[d.id] = d.data().name; });
  log(`Da tai ${snapOI.size} orderItems + ${snapMI.size} menuItems vao cache`);

  log(`May in bep: ${PRINTER_IP}:${PRINTER_PORT}`);
  log('=' .repeat(50));
  log('Dang lang nghe don hang moi...');

  // Lang nghe Firestore real-time (ca don moi lan them mon vao don cu)
  db.collection('bills').onSnapshot(snapshot => {
    snapshot.docChanges().forEach(change => {
      if (change.type === 'added' || change.type === 'modified') {
        handleBillChange(change.doc, change.type).catch(err => log(`Loi xu ly don: ${err.message}`));
      }
    });
  }, err => {
    log(`Loi Firestore: ${err.message}`);
  });
}

main();

File: scripts/kitchen-listener.py
#!/usr/bin/env python3
"""
Kitchen Auto-Print Listener
============================
Chay tren PC tai quan - lang nghe Firestore, tu dong in phieu bep khi co don moi.
Khong can dien thoai goi den print server.

Cai dat:
  pip install firebase-admin python-escpos

Chay:
  python scripts/kitchen-listener.py

Can file serviceAccountKey.json tu Firebase Console:
  Firebase Console -> Project Settings -> Service accounts -> Generate new private key
  Luu file vao: scripts/serviceAccountKey.json
"""

import json
import time
import threading
from datetime import datetime, timezone
from pathlib import Path

import firebase_admin
from firebase_admin import credentials, firestore
from escpos.printer import Network

# ── Cau hinh ──────────────────────────────────────────────────────────────────
PRINTER_IP          = "192.168.1.234"
SERVICE_ACCOUNT_KEY = Path(__file__).parent / "serviceAccountKey.json"

# Dat True neu may in in ra ky tu loi (khong ho tro UTF-8)
USE_ASCII_FALLBACK = False

LINE_WIDTH = 48

# ── Helper ────────────────────────────────────────────────────────────────────
def log(msg):
    t = datetime.now().strftime("%H:%M:%S")
    print(f"[{t}] {msg}", flush=True)

def to_ascii(text):
    import unicodedata
    text = text.replace("d\u0111", "d").replace("\u0110", "D")
    return ''.join(
        c for c in unicodedata.normalize('NFD', text)
        if unicodedata.category(c) != 'Mn'
    )

def t(text):
    return to_ascii(str(text)) if USE_ASCII_FALLBACK else str(text)

def fmt_time(ts):
    """Chuyen Firestore timestamp thanh chuoi gio:phut ngay/thang"""
    if ts is None:
        return datetime.now().strftime("%H:%M %d/%m")
    try:
        dt = ts.astimezone()
        return dt.strftime("%H:%M %d/%m")
    except Exception:
        return datetime.now().strftime("%H:%M %d/%m")

# ── In phieu bep ──────────────────────────────────────────────────────────────
def print_kitchen_ticket(table_label, order_time, item_name, quantity, note=""):
    """In 1 phieu cho 1 mon an"""
    try:
        p = Network(PRINTER_IP, timeout=5)

        p.set(align="center", bold=False, double_height=False, double_width=False)
        p.textln("=" * LINE_WIDTH)

        # Thoi gian
        p.set(align="center")
        p.textln(t(order_time))

        # Ten ban - chu to
        p.set(align="center", bold=True, double_height=True, double_width=True)
        p.textln(t(table_label))
        p.set(bold=False, double_height=False, double_width=False)

        p.set(align="left")
        p.textln("-" * LINE_WIDTH)

        # Ten mon - chu cao
        p.set(align="left", bold=True, double_height=True)
        p.textln(t(item_name))
        p.set(bold=False, double_height=False)

        # So luong - chu to, can phai
        p.set(align="right", bold=True, double_height=True, double_width=True)
        p.textln(f"SL: {quantity}")
        p.set(bold=False, double_height=False, double_width=False)

        if note:
            p.set(align="left")
            p.textln("-" * LINE_WIDTH)
            p.textln(t(f"Ghi chu: {note}"))

        p.set(align="left")
        p.textln("=" * LINE_WIDTH)
        p.ln(2)
        p.cut()
        p.close()
        return True
    except Exception as e:
        log(f"  [LOI IN] {e}")
        return False

# ── Lay ten mon tu orderItems cache ───────────────────────────────────────────
_order_items_cache = {}

def load_order_items(db):
    """Tai toan bo orderItems vao cache"""
    global _order_items_cache
    docs = db.collection("orderItems").stream()
    _order_items_cache = {doc.id: doc.to_dict() for doc in docs}
    log(f"Da tai {len(_order_items_cache)} orderItems vao cache")

def get_item_name(order_item_id, db):
    if order_item_id in _order_items_cache:
        return _order_items_cache[order_item_id].get("name", "Mon khac")
    # Thu doc truc tiep neu chua co trong cache
    doc = db.collection("orderItems").document(order_item_id).get()
    if doc.exists:
        data = doc.to_dict()
        _order_items_cache[order_item_id] = data
        return data.get("name", "Mon khac")
    return "Mon khac"

# ── Xu ly don moi ─────────────────────────────────────────────────────────────
_printed_bills = set()  # Luu cac bill da in de tranh in lai

def handle_new_bill(doc_snapshot, db):
    """Goi khi co bill moi hoac bill duoc cap nhat"""
    bill_id = doc_snapshot.id
    data    = doc_snapshot.to_dict()

    if not data:
        return

    # Chi xu ly don moi tao (khong xu ly cap nhat thanh toan v.v.)
    if bill_id in _printed_bills:
        return

    status = data.get("status", "")
    if status not in ("pending",):
        return

    # Kiem tra bill nay moi tao chua (trong vong 30 giay)
    created_at = data.get("createdAt")
    if created_at:
        try:
            now_utc = datetime.now(timezone.utc)
            bill_time = created_at.astimezone(timezone.utc)
            age_seconds = (now_utc - bill_time).total_seconds()
            if age_seconds > 30:
                # Bill cu (truoc khi script chay) - bo qua
                _printed_bills.add(bill_id)
                return
        except Exception:
            pass

    _printed_bills.add(bill_id)

    # Xac dinh ten ban
    is_takeaway     = data.get("isTakeaway", False)
    takeaway_number = data.get("takeawayNumber")
    table_number    = data.get("tableNumber", "?")
    if is_takeaway and takeaway_number:
        table_label = f"Mang ve #{takeaway_number}"
    else:
        table_label = f"Ban {table_number}"

    order_time = fmt_time(data.get("createdAt"))
    items      = data.get("items", [])

    log(f"Don moi: {table_label} - {len(items)} mon")

    # In tung mon
    success_count = 0
    for item in items:
        order_item_id = item.get("orderItemId")
        quantity      = item.get("quantity", 1)
        note          = item.get("note", "")

        if order_item_id:
            item_name = get_item_name(order_item_id, db)
        else:
            item_name = item.get("customDescription", "Mon khac")

        ok = print_kitchen_ticket(table_label, order_time, item_name, quantity, note)
        if ok:
            success_count += 1
            log(f"  [{success_count}] {item_name} x{quantity} - OK")
        else:
            log(f"  LOI khi in: {item_name} x{quantity}")

    log(f"Xong: {table_label} - {success_count}/{len(items)} phieu in thanh cong")

# ── Main ──────────────────────────────────────────────────────────────────────
def main():
    # Kiem tra service account key
    if not SERVICE_ACCOUNT_KEY.exists():
        print(f"""
[LOI] Khong tim thay file: {SERVICE_ACCOUNT_KEY}

Huong dan lay service account key:
  1. Vao Firebase Console: https://console.firebase.google.com
  2. Chon project CaMauQuan
  3. Project Settings (bieu tuong rang cua) -> Service accounts
  4. Nhan "Generate new private key" -> tai file JSON
  5. Doi ten thanh: serviceAccountKey.json
  6. Dat vao thu muc: {SERVICE_ACCOUNT_KEY.parent}
""")
        return

    # Ket noi Firebase
    log("Dang ket noi Firebase...")
    cred = credentials.Certificate(str(SERVICE_ACCOUNT_KEY))
    firebase_admin.initialize_app(cred)
    db = firestore.client()
    log("Ket noi Firebase thanh cong")

    # Tai orderItems cache
    load_order_items(db)

    # Lang nghe bills collection
    log(f"Bat dau lang nghe don hang moi...")
    log(f"May in: {PRINTER_IP}")
    log("=" * 50)

    def on_snapshot(col_snapshot, changes, read_time):
        for change in changes:
            if change.type.name == "ADDED":
                try:
                    handle_new_bill(change.document, db)
                except Exception as e:
                    log(f"Loi xu ly don: {e}")

    # Dang ky listener tren collection bills
    col_ref = db.collection("bills")
    col_ref.on_snapshot(on_snapshot)

    # Giu script chay mai mai
    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        log("Da dung lang nghe.")

if __name__ == "__main__":
    main()
